doc/tooltip-api.txt

```diff
@@ -0,0 +1,172 @@
+
+    Returns the window ID of the created popup.
+
+show_pum_tooltip({lines}, {filetype}, {syn-ranges}, {options})
+
+    Shows a tooltip associated with the currently selected item in the popup
+    menu. It can be either to the left/right of the item, depending on where
+    there is enough space. Settings from |'completepopup'| is taken into
+    account.
+
+    The preview's content is set depending on {lines}, {syn-ranges} and
+    {filetype}, see above.
+
+    Returns the window ID of the created popup.
+
+update_pum_tooltip({winid}, {lines}, {filetype}, {syn-ranges}, {options})
+
+    Changes the content of the tooltip associated with the currently selected
+    item in the popup menu. {winid} must be the value returned by
+    `show_pum_tooltip`.
+
+
+In addition, there are utility functions useful for interacting with floats:
+
+parse_lsp_response({data})
+
+    Parses a LSP response (be it a String, MarkedString, Markdown segment,
+    MarkupContent, ...; from hereon just {data}) to 
+        
+        - a |list| of |String|s that represent the "stripped" content, e.g.
+          markdown `*bold*` strings will be converted to just `bold`, code
+          blocks will have the markers removed, etc. It's the "plain-text, no
+          colour" version of the {data};
+
+        - a |list| of |Dicts| that specifies the |syn-ranges|, i.e., of 
+          character ranges and Vim syntax groups to apply to them. 
+
+    Returns a tuple |[lines,syn-ranges]| .
+
+TODO: move to `util.vim`?
+
+
+==============================================================================
+INTERMEDIATE-LEVEL API                                *intermediate-level-api*
+
+These functions abstract the low-level backend-specific API for interacting
+with tooltip-type windows. They have a common signature but are implemented
+separately for each backend.
+
+Currently, the following backends are implemented:
+
+1. `popup` windows for vim 8.1.1517+
+2. `floating windows` for neovim 0.4.0+
+3. TODO: `preview` buffers as a fallback 
+
+create_tooltip()
+
+    Creates a hidden floating window, with undefined position, and an empty
+    buffer associated with it.
+
+    Returns the created window ID.
+
+
+set_tooltip_contents({winid}, {lines})
+
+    Updates the contents of a given floating window, and unhides it. Also
+    retriggers size calculation.
+
+
+set_tooltip_position({winid}, {options})
+
+    Sets the position of the given floating window, be it cursor-relative,
+    pum-relative, etc. Also retriggers size calculation.
+
+    TODO: Which {options}? width in characters or bytes? 0- or 1-indexed?
+
+
+close_tooltip({winid})
+
+    Closes the tooltip with the ID {winid}.
+
+
+In addition, there are common utility functions:
+
+get_size_info({lines})
+     
+     Get width and height needed to show the line-wrapped contents of a list
+     of strings.
+
+     Returns a tuple of |[lines,width]| 
+
+     TODO: in characters or bytes? 0- or 1-indexed?
```

Comment:
0 and 1 indexing doesn't make much sense for sizes ;)

Reply:
True :) I was thinking here of utility functions in general, such as ones that compute position (for moving tooltips away from the buffer boundary).